### PR TITLE
Azure.Provisioning: Merge BicepValue into Models

### LINKED ISSUE #46557: Azure.Provisioning: Remove `BicepValue<T>` for models
### Library name

Azure.Provisioning

### Please describe the feature.

Using `BicepValue<T>` for model types results in a lot of unnatural `foo.Value!.Bar.Value!.Baz`.  Models deriving from `ProvisioningConstruct` should encapsulate the `T | Expression` concept so we only have to write `foo.Bar.Baz`.

### PR DESCRIPTION
**Pull request by tg-msft**
Fixes #46557

(Leaving in a draft w/o generated code for easier initial review)

### CODE REVIEW

File: sdk/provisioning/Azure.Provisioning/src/BicepValue.cs [hunk 1]
```diff
@@ -2,110 +2,131 @@
...
-// models and enums out (kind of like we do with ResourceReference already) so
-// we get a better intellisense experience.  That'll also remove a lot of the
-// clutter in this class hierarchy.
-
 /// <summary>
 /// Represents the value of a property that could be a literal .NET value, a
 /// Bicep expression, or it could be unset.
 /// </summary>
-public abstract class BicepValue
+public abstract class BicepValue : IBicepValue
```
Comment (christothes) [inline sdk/provisioning/Azure.Provisioning/src/BicepValue.cs:15]: Just curious - why the change to an interface?

File: sdk/provisioning/Azure.Provisioning/src/IBicepValue.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+using Azure.Provisioning.Expressions;
+using Azure.Provisioning.Primitives;
+
+namespace Azure.Provisioning;
+
+/// <summary>
+/// Represents the value of a property that could be a literal .NET value, a
+/// Bicep expression, or it could be unset.
+/// </summary>
+public interface IBicepValue
```
Comment (christothes) [inline sdk/provisioning/Azure.Provisioning/src/IBicepValue.cs:13]: nit: The docs are inconsistent with language like "Gets/Get/Tracks..." Should we be consistent with "Gets..." or "Gets or sets.."?

File: sdk/provisioning/Azure.Provisioning/src/Primitives/BicepValueReference.cs [hunk 3]
```diff
@@ -36,6 +37,5 @@ internal BicepExpression GetReference()
         return target;
     }
 
-    public override string ToString() =>
-        $"<{GetReference()} from {Construct.GetType().Name}.{PropertyName}>";
+    public override string ToString() => GetReference().ToString();
```
Comment (christothes) [inline sdk/provisioning/Azure.Provisioning/src/Primitives/BicepValueReference.cs:40]: Is it ok that this throws when the construct is not named?

File: sdk/provisioning/Azure.Provisioning/src/Primitives/ProvisionableResource.cs [hunk 4]
```diff
@@ -33,7 +33,22 @@ public abstract class ProvisionableResource(string bicepIdentifier, ResourceType
...
     /// a new resource.
     /// </summary>
-    public bool IsExistingResource { get; protected set; } = false;
+    public bool IsExistingResource
+    {
+        get => _isExisting;
+        protected set
+        {
+            _isExisting = value;
+            if (_isExisting)
```
Comment (christothes) [inline sdk/provisioning/Azure.Provisioning/src/Primitives/ProvisionableResource.cs:42]: Possible threading bug?
```suggestion
            _isExisting = value;
            if (value)
```

File: sdk/provisioning/Generator/src/Model/Resource.cs [hunk 5]
```diff
@@ -127,12 +146,36 @@ public override void Generate()
                         writer.Write($" ?? \"{DefaultResourceVersion}\"");
                     }
                     writer.WriteLine(")");
+                    using (writer.Scope("{", "}")) { }
```
Comment (christothes) [inline sdk/provisioning/Generator/src/Model/Resource.cs:149]: is this intentionally empty?

File: sdk/provisioning/Generator/src/Model/SimpleModel.cs [hunk 6]
```diff
@@ -88,16 +103,41 @@ public override void Generate()
                         writer.Write(" : base()");
                     }
                     writer.WriteLine();
+                    using (writer.Scope("{", "}")) { }
```
Comment (christothes) [inline sdk/provisioning/Generator/src/Model/SimpleModel.cs:106]: intentionally empty?

File: sdk/provisioning/Azure.Provisioning/src/BicepValue.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/src/BicepValue.cs:15]: There used to be two "hierarchies":
```mermaid
graph LR
  A[Provisonable] --> Infrastructure
  A --> B[ProvisionableConstruct]
  B --> C[NamedProvisionableConstruct]
  C --> D[Variable]
  C --> ProvisionableResource
  
  
  X[BicepValue] --> BicepValueOfT
  X --> BicepList
  X --> BicepDict 
```

This PR bridges them:

```mermaid
graph LR
  A[Provisonable] --> Infrastructure
  A --> B[ProvisionableConstruct]
  B --> C[NamedProvisionableConstruct]
  C --> D[Variable]
  C --> ProvisionableResource
  
  Z[IBicepValue] --> B
  Z --> X
    
  X[BicepValue] --> BicepValueOfT
  X --> BicepList
  X --> BicepDict 
```

The net result is now all our model types are implicitly `T | Expression` without being wrapped in `BicepValue<T>`.  This means we can replace code like `storage.PrimaryEndpoints.Value!.BlobUri` with just `storage.PrimaryEndpoints.BlobUri` because the `StorageAccountEndpoints` now knows whether it has a literal value or expression and just manages it for us.  When you start looking at ACA examples, the API was annoyingly shouty with `properties.Foo.Value!.Bar.Value!.Baz = 3` and this will make a big difference to the overall feel.

File: sdk/provisioning/Azure.Provisioning/src/Primitives/BicepValueReference.cs (hunk 3 above)
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/src/Primitives/BicepValueReference.cs:40]: I was debating this.  I think I'm the only person who will ever look at this type, but I'll patch this to be friendlier.

File: sdk/provisioning/Azure.Provisioning/src/Primitives/ProvisionableResource.cs (hunk 4 above)
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/src/Primitives/ProvisionableResource.cs:42]: None of this is designed to be thread safe.  You'll have to manage synchronization yourself if you want to share these objects across threads (i.e., just like our Azure SDK model types).

File: sdk/provisioning/Azure.Provisioning/tests/ProvisioningTestBase.cs [hunk 7]
```diff
@@ -20,7 +20,7 @@
 namespace Azure.Provisioning.Tests;
 
 [AsyncOnly]
-[LiveOnly] // Ignore tests in the CI for now
+// [LiveOnly] // Ignore tests in the CI for now
```
[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/provisioning/Azure.Provisioning/tests/ProvisioningTestBase.cs:23]: Nope.  Thanks!

File: sdk/provisioning/Generator/src/Model/Resource.cs (hunk 5 above)
Comment (tg-msft) [reply sdk/provisioning/Generator/src/Model/Resource.cs:149]: Yes, this is generating an empty .ctor for us.  In the past we were initializing all the nested members immediately and relying on them being wrapped in `BicepValue<T>` to avoid loading THE ENTIRE OBJECT GRAPH.  We've had to shuffle things around to prevent that and now there's an `Initialize` method that we call before any property is touched.  Let me add an example resource and example model to the PR to help show the generated code diffs.

File: sdk/provisioning/Generator/src/Model/SimpleModel.cs (hunk 6 above)
Comment (tg-msft) [reply sdk/provisioning/Generator/src/Model/SimpleModel.cs:106]: Yes - ditto of above, but for models.

File: sdk/provisioning/Azure.Provisioning.AppService/tests/BasicAppServiceTests.cs [hunk 8]
```diff
@@ -81,12 +82,12 @@ await test.Define(
                                     new AppServiceNameValuePair
                                     {
                                         Name = "AzureWebJobsStorage",
-                                        Value = BicepFunction.Interpolate($"DefaultEndpointsProtocol=https;AccountName={storage.Name};EndpointSuffix=core.windows.net;AccountKey={storage.GetKeys()[0].Value!.Value}")
+                                        Value = BicepFunction.Interpolate($"DefaultEndpointsProtocol=https;AccountName={storage.Name};EndpointSuffix=core.windows.net;AccountKey={storage.GetKeys()[0].Unwrap().Value}")
```
Comment (tg-msft) [inline sdk/provisioning/Azure.Provisioning.AppService/tests/BasicAppServiceTests.cs:85]: This is an example of where things break down a little.  When you have a collection of models, things still get wrapped in a `BicepValue<T>` (i.e., `public BicepList<StorageAccountKey> GetKeys()` is the signature which means `GetKeys()[0]` is a `BicepValue<StorageAccountKey>`).  `Unwrap` turns `BicepValue<StorageAccountKey>` into an identical `StorageAccountKey` where everything's easier to access again.  These aren't super common, so this tradeoff was less awful than duplicating collections for models.

File: sdk/provisioning/Azure.Provisioning.KeyVault/tests/BasicKeyVaultTests.cs [hunk 9]
```diff
@@ -97,6 +97,7 @@ await test.Define(
 
                 infra.Add(new ProvisioningOutput("name", typeof(string)) { Value = kv.Name });
                 infra.Add(new ProvisioningOutput("resourceId", typeof(string)) { Value = kv.Id });
+                infra.Add(new ProvisioningOutput("vaultUri", typeof(string)) { Value = kv.Properties.VaultUri });
```
Comment (tg-msft) [inline sdk/provisioning/Azure.Provisioning.KeyVault/tests/BasicKeyVaultTests.cs:100]: Dotting down multiple levels wasn't possible previously but now works fine

File: sdk/provisioning/Azure.Provisioning.Storage/tests/BasicStorageTests.cs [hunk 10]
```diff
@@ -23,7 +23,7 @@ await test.Define(
             new StorageAccount("storage", StorageAccount.ResourceVersions.V2023_01_01)
             {
                 Kind = StorageKind.StorageV2,
-                Sku = new StorageSku { Name = StorageSkuName.StandardLrs },
+                Sku = { Name = StorageSkuName.StandardLrs },
```
Comment (tg-msft) [inline sdk/provisioning/Azure.Provisioning.Storage/tests/BasicStorageTests.cs:26]: Now we'll lazily initialize models for you so you can just `storage.Sku.Name = ...` if you want without a `new StorageSku();` first.

File: sdk/provisioning/Azure.Provisioning.Storage/tests/BasicStorageTests.cs [hunk 11]
```diff
@@ -267,7 +267,7 @@ await test.Define(
                 BlobService blobs = new(nameof(blobs)) { Parent = storage };
                 infra.Add(blobs);
 
-                infra.Add(new ProvisioningOutput("blobs_endpoint", typeof(string)) { Value = storage.PrimaryEndpoints.Value!.BlobUri });
+                infra.Add(new ProvisioningOutput("blobs_endpoint", typeof(string)) { Value = storage.PrimaryEndpoints.BlobUri });
```
Comment (tg-msft) [inline sdk/provisioning/Azure.Provisioning.Storage/tests/BasicStorageTests.cs:270]: This was the ugliest remaining part of the API and almost crippled the dev experience in some scenarios.

**Review by tg-msft — COMMENTED** [review]
Pointing out a few impacts of these changes

**Review by christothes — APPROVED** [review]
LGTM
